Description: fix IO safety trip up during tests
 When the files opened for the tests are opened in rust_tests and their
 FDs are dropped, the IO safety checks will notice their FDs have been
 closed by the test invocation and panic at runtime. Prevent this by
 opening raw fds directly.
Author: Ananthu C V <weepingclown@disroot.org>
Bug: https://github.com/cptpcrd/close_fds/issues/1
Forwarded: https://github.com/cptpcrd/close_fds/pull/2
Last-Update: 2024-11-22

--- a/tests/test_close_fds.rs
+++ b/tests/test_close_fds.rs
@@ -57,11 +57,10 @@
 ) {
-    let f1 = std::fs::File::open("/").unwrap();
-    let f2 = std::fs::File::open("/").unwrap();
-    let f3 = std::fs::File::open("/").unwrap();
 
-    let fd1 = f1.as_raw_fd();
-    let fd2 = f2.as_raw_fd();
-    let fd3 = f3.as_raw_fd();
+    let path = std::ffi::CString::new("/").unwrap();
 
-    drop(f3);
+    let fd1 = unsafe { libc::open(path.as_ptr(), libc::O_RDONLY) };
+    let fd2 = unsafe { libc::open(path.as_ptr(), libc::O_RDONLY) };
+    let fd3 = unsafe { libc::open(path.as_ptr(), libc::O_RDONLY) };
+
+    unsafe { libc::close(fd3) };
 
